backend-rust/src/indexer.rs

```diff
@@ -2554,13 +2570,39 @@ impl PreparedContractUpdate {
     }
 }
 
-struct PreparedContractUpgrade {
+enum PreparedContractTraceEvent {
+    /// Potential module link events from a smart contract upgrade
+    Upgrade(PreparedTraceEventUpgrade),
+    /// Transfer to account.
+    Transfer(PreparedTraceEventTransfer),
+    /// Send messages (and CCD) updating another contract.
+    Update(PreparedTraceEventUpdate),
+    /// Nothing further needs to be tracked.
+    NoEvent,
+}
+
+impl PreparedContractTraceEvent {
+    async fn save(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+        transaction_index: i64,
+    ) -> anyhow::Result<()> {
+        match self {
+            PreparedContractTraceEvent::Upgrade(event) => event.save(tx, transaction_index).await,
```

Comment:
Please ensure this here makes sense. I have limited domain knowledge on this particular field about expected behaviour in the underlying types.